core/common/src/main/java/alluxio/AbstractClient.java

```diff
@@ -196,14 +196,11 @@ public synchronized void connect() throws IOException, ConnectionFailedException
   public synchronized void disconnect() {
     if (mConnected) {
       LOG.debug("Disconnecting from the {} {} {}", getServiceName(), mMode, mAddress);
-      mConnected = false;
-    }
-    try {
       beforeDisconnect();
       if (mProtocol != null) {
```

Comment:
I agree that we shouldn't mask an invariant violation - could we make it clearer with a `Preconditions.checkNotNull`?